File: app/src/models/watchtime.model.ts
export interface WatchtimeDbModel {
    watchtime_id: number;
    episode_id: number;
    percentage_watched: number;
    stopped_time: number;
    tenant_id: string;
}

export interface WatchtimeModel extends WatchtimeDbModel {
    clone(): WatchtimeModel;
}


/**
 * Used for API only
 */
export interface TotalProgressionModel {
    total_progression: number;
}

/**
 * Used for API only
 */
export interface WatchtimeCreateModel {
    episode_id: number;
    percentage_watched: number;
    stopped_time: number;
}

/**
 * Used for API only
 */
export interface WatchtimeUpdateModel {
    percentage_watched: number | null;
    stopped_time: number | null;
}

export function WatchtimeModel(episode_id: number, percentage_watched: number, stopped_time: number, tenant_id: string): WatchtimeModel;
export function WatchtimeModel(watchtime_id: number, episode_id: number, percentage_watched: number, stopped_time: number, tenant_id: string): WatchtimeModel;

export function WatchtimeModel(...args: unknown[]): WatchtimeModel {
    if (args.length == 4) {
        args.unshift(0);
    }

    return _WatchtimeModel(...args as [number, number, number, number, string]);
}

function _WatchtimeModel(watchtime_id: number, episode_id: number, percentage_watched: number, stopped_time: number, tenant_id: string): WatchtimeModel {
    const obj: WatchtimeModel = {
        watchtime_id: watchtime_id,
        episode_id: episode_id,
        percentage_watched: percentage_watched,
        stopped_time: stopped_time,
        tenant_id: tenant_id,
        clone: clone,
    };

    obj.clone = obj.clone.bind(obj);
    return obj;
}

function clone(this: WatchtimeModel): WatchtimeModel {
    const obj: WatchtimeModel = {
        watchtime_id: this.watchtime_id,
        episode_id: this.episode_id,
        percentage_watched: this.percentage_watched,
        stopped_time: this.stopped_time,
        tenant_id: this.tenant_id,
        clone: clone,
    };

    obj.clone = obj.clone.bind(obj);
    return obj;
}

File: server/AniStream/DTO/WatchTimeModel.cs
namespace AniStream.API.DTO;

public sealed class WatchTimeModel
{
    public required int WatchtimeId { get; set; }

    public required int EpisodeId { get; set; }

    public required int PercentageWatched { get; set; }

    public required double StoppedTime { get; set; }

    public required string TenantId { get; set; }
}

public sealed class WatchTimeCreateModel
{
    public required int EpisodeId { get; set; }

    public required int PercentageWatched { get; set; }

    public required double StoppedTime { get; set; }
}

public sealed class WatchTimeUpdateModel
{
    public required int? PercentageWatched { get; set; }

    public required double? StoppedTime { get; set; }
}


public sealed class WatchTimeTotalModel
{
    public required int TotalProgression { get; set; }
}

internal static class WatchTimeModelHelper
{
    public static WatchTimeModel ToDTO(this Models.WatchTimeModel model)
    {
        return new WatchTimeModel
        {
            WatchtimeId = model.WatchtimeId,
            EpisodeId = model.EpisodeId,
            PercentageWatched = model.PercentageWatched,
            StoppedTime = model.StoppedTime,
            TenantId = model.TenantId
        };
    }
}